### PR TITLE
[Cleanup] Partial cleanup & refactor of different parts of the codebase

### PR DESCRIPTION
**Pull request by Olshansk**
This PR does not introduce any functional changes to the logic.

- **Tests**: It is a partial cleanup & refatctor of functions, names, options and all tests remain the same & passing.
- **Why**: Make it easier for new developers to understand the SMT
- **Up next**:
	- Continud cleanup & refactor
	- Adding relay counts (not just compute units) to the tree (needed for relay mining)

### CODE REVIEW

File: docs/smt.md
Comment (h5law) [inline docs/smt.md:14]: You really don't like british spelling ahah

File: docs/smt.md
Comment (h5law) [inline docs/smt.md:64]:
```suggestion
prefix, path and value.
```

File: docs/smt.md
Comment (h5law) [inline docs/smt.md:89]:
```suggestion
They are used to represent a common path in the trie and as such
```

File: extension_node.go
Comment (h5law) [inline extension_node.go:134]:
```suggestion
// expand returns the inner node that represents the end of the singly
```

File: hasher.go
Comment (h5law) [inline hasher.go:50]:
```suggestion
// nilPathHasher is a dummy hasher that returns its input - it should not be used outside of the closest proof verification logic
type nilPathHasher struct {
```

File: hasher.go
Comment (h5law) [inline hasher.go:61]: I don't think this should be exported AFAIK the validation logic for the ClosestProof automatically handles this.

File: hasher.go
Comment (h5law) [inline hasher.go:114]: Either add comments to all of these or none `digestLeafNode` sets the example but the names are self explanatory - your call

File: node_encoders.go
Comment (h5law) [inline node_encoders.go:11]: Protobuf encoding and decoding could potentially slow down the tree - keeping with go types would be my preffered way of doing things IMO. Gob encoding is optimised for go types anyway

File: options.go
Comment (h5law) [inline options.go:26]:
```suggestion
func noHasherSpec(hasher hash.Hash, sumTrie bool) *TrieSpec {
```
Do not export this, this is only needed in the validation logic for closest proofs and **should never** be used for a trie

File: proofs.go
Comment (h5law) [inline proofs.go:369]: https://github.com/pokt-network/smt/blob/fea9ecbf697600acaa87b4903e72901c98fc286f/proofs.go#L340

See how closest proofs are handled here to understand why we dont need NilPathHasher exposed - your branch is also out of sync

File: proofs_test.go
Comment (h5law) [inline proofs_test.go:31]: Again with the `z`s lmao

File: root_test.go
Comment (h5law) [inline root_test.go:61]: It would be nice to require the tree sum be 45 here instead

Comment (h5law) [inline root_test.go:61]:
```suggestion
				require.Equal(t, 45, trie.Sum())
```

File: smst.go
Comment (h5law) [inline smst.go:36]: This is because the SMST pre-hashes its paths and thus would result in double path hashing as it is a wrapper around the SMT. The SMST uses its own path hashing logic to account for the sum so the SMT underneath needs a nil path hasher and just uses the paths provided.

File: smst.go
Comment (h5law) [inline smst.go:151]:
```suggestion
	rootDigest := smst.Root()
	if !smst.Spec().sumTrie {
		panic("SMST: not a merkle sum trie")
	}
```

File: smt.go
Comment (h5law) [inline smt.go:607]: I feel like we can introduce a helper as this logic is duplicated

File: types.go
Comment (h5law) [inline types.go:4]: digest is the correct term for the output of a hashing function IIRC

File: types.go
Comment (h5law) [inline types.go:5]: Sum() would have to be no-op but could be done

File: types.go
Comment (h5law) [inline types.go:6]: No because what if this is an in memory trie?

**Review by h5law — CHANGES_REQUESTED** [review]
Overall great refactor - bring the branch in line with main though. Left some comments with context and suggestions.

Nice job cleaning up my baby - but I miss my `s`s over your `z`s

File: docs/smt.md
Comment (Olshansk) [reply docs/smt.md:14]: 🇺🇸

File: options.go
Comment (Olshansk) [reply options.go:26]: Removed!

File: smt.go
Comment (Olshansk) [reply smt.go:607]: Refactored to use helpers but the logic is unique between a sum & non-sum.

**Comment by Olshansk** [comment]
@h5law Appreciate the quick review!

**Summary:**
1. The branch was in sync with `main` so not sure about that part.
2. Updated comments & left TODOs based on some of your comments
3. 🇺🇸 

**Next steps:**
1. Note that this is one (of several) refactors. I think there's more work to be done so it's fully accessible to all developers.
2. I'm going to be adding this: `Adding relay counts (i.e. the number of leafs) to the tree for relay mining`. I have some ideas on how to do this. If you have a strong opinion of how to go about it, lmk. If not, I'll tackle it.

**Comment by h5law** [comment]
@Olshansk 

> The branch was in sync with main so not sure about that part

Your VerifyClosestProof method was old so it seems out of sync maybe a rebase is in order ahah gl

> Adding relay counts (i.e. the number of leafs) to the tree for relay mining

I will tackle this if you give me a spec/idea of the desired results and deadline

**Comment by Olshansk** [comment]
> Your VerifyClosestProof method was old so it seems out of sync maybe a rebase is in order ahah gl

The header on main is: `func VerifyClosestProof(proof *SparseMerkleClosestProof, root []byte, spec *TrieSpec) (bool, error) {`

Screenshot 2024-06-04 at 2 26 26 PM

The header in this PR is the same:

Screenshot 2024-06-04 at 2 27 19 PM

What am I missing?

**Comment by Olshansk** [comment]
> I will tackle this if you give me a spec/idea of the desired results and deadline

Noted. Don't want to take too much of your time so I'll handle it.

**Comment by Olshansk** [comment]
@h5law Please see https://github.com/pokt-network/smt/pull/44#issuecomment-2148444978

**Comment by h5law** [comment]
> Noted. Don't want to take too much of your time so I'll handle it.

I would kinda like to do it ngl ahaha but up to you

Regarding the out of sync I think I was in the error there ill do a once over now and see if I missed anything otherwise approve

File: smt.go
Comment (h5law) [inline smt.go:207]:
```suggestion
	// The node must be an innerNode so depending on which side of the
	// branch the newLeaf should be call update() again on that node in
	// order to correctly place the newLeaf
	inner := node.(*innerNode)
```

File: types.go
Comment (h5law) [inline types.go:10]: Not sure where you moved it but itd now called `sumSizeBits` shouldnt it be `sumSizeBytes` as it represents a uint64 not uint8

**Review by h5law — APPROVED** [review]
Overall LGTM, approving now as the majority of my comments are in fact comments themselves and some naming - easy fixes no need for another review IMO. Great job on the refactor 🚀

**Pull request merged**
